statistics/sample_serial_test.go

```diff
@@ -0,0 +1,328 @@
+			ColsFieldType:   []*types.FieldType{types.NewFieldType(mysql.TypeLonglong), types.NewFieldType(mysql.TypeLonglong)},
+		}
+		require.Nil(t, s.rs.Close())
+		collectors, pkBuilder, err := builder.CollectColumnStats()
+		require.NoError(t, err)
+		require.Nil(t, pkBuilder)
+		for _, collector := range collectors {
+			p := SampleCollectorToProto(collector)
+			s := SampleCollectorFromProto(p)
+			require.Equal(t, s.Count, collector.Count)
+			require.Equal(t, s.NullCount, collector.NullCount)
+			require.Equal(t, s.CMSketch.TotalCount(), collector.CMSketch.TotalCount())
+			require.Equal(t, s.FMSketch.NDV(), collector.FMSketch.NDV())
+			require.Equal(t, s.TotalSize, collector.TotalSize)
+			require.Equal(t, len(s.Samples), len(collector.Samples))
+		}
+	}
+}
+
+func (s *testSampleSuite) recordSetForWeightSamplingTest(size int) *recordSet {
+	r := &recordSet{
+		data:  make([]types.Datum, 0, size),
+		count: size,
+	}
+	for i := 0; i < size; i++ {
+		r.data = append(r.data, types.NewIntDatum(int64(i)))
+	}
+	r.setFields(mysql.TypeLonglong)
+	return r
+}
+
+func (s *testSampleSuite) recordSetForDistributedSamplingTest(size, batch int) []*recordSet {
+	sets := make([]*recordSet, 0, batch)
+	batchSize := size / batch
+	for i := 0; i < batch; i++ {
+		r := &recordSet{
+			data:  make([]types.Datum, 0, batchSize),
+			count: batchSize,
+		}
+		for j := 0; j < size/batch; j++ {
+			r.data = append(r.data, types.NewIntDatum(int64(j+batchSize*i)))
+		}
+		r.setFields(mysql.TypeLonglong)
+		sets = append(sets, r)
+	}
+	return sets
+}
+
+func SubTestWeightedSampling(s *testSampleSuite) func(*testing.T) {
+	return func(t *testing.T) {
+		sampleNum := int64(20)
+		rowNum := 100
+		loopCnt := 1000
+		rs := s.recordSetForWeightSamplingTest(rowNum)
+		sc := mock.NewContext().GetSessionVars().StmtCtx
+		// The loop which is commented out is used for stability test.
+		// This test can run 800 times in a row without any failure.
+		// for x := 0; x < 800; x++ {
+		itemCnt := make([]int, rowNum)
+		for loopI := 0; loopI < loopCnt; loopI++ {
+			builder := &ReservoirRowSampleBuilder{
+				Sc:              sc,
+				RecordSet:       rs,
+				ColsFieldType:   []*types.FieldType{types.NewFieldType(mysql.TypeLonglong)},
+				Collators:       make([]collate.Collator, 1),
+				ColGroups:       nil,
+				MaxSampleSize:   int(sampleNum),
+				MaxFMSketchSize: 1000,
+				Rng:             rand.New(rand.NewSource(time.Now().UnixNano())),
+			}
+			collector, err := builder.Collect()
+			require.NoError(t, err)
+			for i := 0; i < collector.MaxSampleSize; i++ {
+				a := collector.Samples[i].Columns[0].GetInt64()
+				itemCnt[a]++
+			}
+			require.Nil(t, rs.Close())
+		}
+		expFrequency := float64(sampleNum) * float64(loopCnt) / float64(rowNum)
+		delta := 0.5
+		for _, cnt := range itemCnt {
+			if float64(cnt) < expFrequency/(1+delta) || float64(cnt) > expFrequency*(1+delta) {
+				require.Truef(t, false, "The frequency %v is exceed the Chernoff Bound", cnt)
+			}
```

Comment:
```suggestion
			require.Conditionf(t, func() bool { return float64(cnt) >= expFrequency/(1+delta) && float64(cnt) <= expFrequency*(1+delta) }, "The frequency %v is exceed the Chernoff Bound", cnt)
			}
```

Reply:
testify has a matcher named [`InDelta`](https://pkg.go.dev/github.com/stretchr/testify@v1.7.0/require#InDelta) you can make use of.

Reply:
thanks, updated.
as per https://stackoverflow.com/questions/45310329/difference-between-indelta-and-inepsilon

```
InDelta receives an absolute value and checks that the difference is less than that value.
InEpsilon receives a % of difference that would be acceptable.
```

used InEpsilon here.